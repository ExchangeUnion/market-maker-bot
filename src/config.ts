import { DotenvParseOutput } from 'dotenv';
import { Observable, of } from 'rxjs';
import { map, pluck } from 'rxjs/operators';
import { Level } from './logger';

export type Config = {
  LOG_LEVEL: Level;
  CEX: string;
  CEX_API_KEY: string;
  CEX_API_SECRET: string;
  DATA_DIR: string;
  OPENDEX_CERT_PATH: string;
  OPENDEX_RPC_HOST: string;
  OPENDEX_RPC_PORT: string;
  MARGIN: string;
  BASEASSET: string;
  QUOTEASSET: string;
  CEX_BASEASSET: string;
  CEX_QUOTEASSET: string;
  TEST_CENTRALIZED_EXCHANGE_BASEASSET_BALANCE: string;
  TEST_CENTRALIZED_EXCHANGE_QUOTEASSET_BALANCE: string;
  TEST_MODE: boolean;
};

const REQUIRED_CONFIGURATION_OPTIONS = [
  'LOG_LEVEL',
  'CEX',
  'DATA_DIR',
  'OPENDEX_CERT_PATH',
  'OPENDEX_RPC_HOST',
  'OPENDEX_RPC_PORT',
  'MARGIN',
  'BASEASSET',
  'QUOTEASSET',
  'TEST_MODE',
];

const REQUIRED_CONFIGURATION_OPTIONS_TEST_MODE_ENABLED = [
  'TEST_CENTRALIZED_EXCHANGE_BASEASSET_BALANCE',
  'TEST_CENTRALIZED_EXCHANGE_QUOTEASSET_BALANCE',
];

const REQUIRED_CONFIGURATION_OPTIONS_TEST_MODE_DISABLED = [
  'CEX_API_KEY',
  'CEX_API_SECRET',
];

const OPTIONAL_CONFIG = ['CEX_BASEASSET', 'CEX_QUOTEASSET'];

const setLogLevel = (logLevel: string): Level => {
  return Object.values(Level).reduce((finalLevel, level) => {
    if (logLevel === level) {
      return level;
    }
    return finalLevel;
  }, Level.Trace);
};

const getEnvironmentConfig = (): DotenvParseOutput => {
  const environmentConfig = REQUIRED_CONFIGURATION_OPTIONS.concat(
    REQUIRED_CONFIGURATION_OPTIONS_TEST_MODE_ENABLED,
    REQUIRED_CONFIGURATION_OPTIONS_TEST_MODE_DISABLED,
    OPTIONAL_CONFIG
  ).reduce((envConfig: DotenvParseOutput, configOption) => {
    if (process.env[configOption]) {
      return {
        ...envConfig,
        [configOption]: process.env[configOption]!,
      };
    }
    return envConfig;
  }, {});
  return environmentConfig;
};

const getMissingOptions = (config: DotenvParseOutput): string => {
  const ADDITIONAL_CONF_OPTIONS =
    config['TEST_MODE'] === 'true'
      ? REQUIRED_CONFIGURATION_OPTIONS_TEST_MODE_ENABLED
      : REQUIRED_CONFIGURATION_OPTIONS_TEST_MODE_DISABLED;
  return REQUIRED_CONFIGURATION_OPTIONS.concat(ADDITIONAL_CONF_OPTIONS)
    .reduce((missingOptions: string[], configOption) => {
      if (!config[configOption]) {
        return missingOptions.concat(configOption);
      }
      return missingOptions;
    }, [])
    .join(', ');
};

const checkConfigOptions = (dotEnvConfig: DotenvParseOutput): Config => {
  const config = {
    ...dotEnvConfig,
    ...getEnvironmentConfig(),
  };
  const missingOptions = getMissingOptions(config);
  if (missingOptions) {
    throw new Error(
      `Incomplete configuration. Please add the following options to .env or as environment variables: ${missingOptions}`
    );
  }
  const BASEASSET = config.BASEASSET.toUpperCase();
  const QUOTEASSET = config.QUOTEASSET.toUpperCase();
  const CEX_BASEASSET = config.CEX_BASEASSET
    ? config.CEX_BASEASSET.toUpperCase()
    : BASEASSET;
  const CEX_QUOTEASSET = config.CEX_QUOTEASSET
    ? config.CEX_QUOTEASSET.toUpperCase()
    : QUOTEASSET;
  const verifiedConfig = {
    ...config,
    LOG_LEVEL: setLogLevel(config.LOG_LEVEL),
    TEST_MODE: config.TEST_MODE === 'true' ? true : false,
    CEX: config.CEX.toUpperCase(),
    BASEASSET,
    QUOTEASSET,
    CEX_BASEASSET,
    CEX_QUOTEASSET,
  };
  return verifiedConfig as Config;
};

const getConfig$ = (): Observable<Config> => {
  return of(require('dotenv').config()).pipe(
    pluck('parsed'),
    map(checkConfigOptions)
  );
};

export { getConfig$, checkConfigOptions };
