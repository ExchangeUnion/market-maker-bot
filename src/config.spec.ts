import { checkConfigOptions } from './config';

describe('checkConfigOptions', () => {
  describe('TEST_MODE enabled', () => {
    const validLiveCEXdisabledConf = {
      LOG_LEVEL: 'trace',
      CEX: 'binance',
      DATA_DIR: '/some/data/path',
      OPENDEX_CERT_PATH: '/some/cert/path',
      OPENDEX_RPC_HOST: 'localhost',
      OPENDEX_RPC_PORT: '1234',
      MARGIN: '0.06',
      BASEASSET: 'btc',
      QUOTEASSET: 'usdt',
      TEST_CENTRALIZED_EXCHANGE_BASEASSET_BALANCE: '321',
      TEST_CENTRALIZED_EXCHANGE_QUOTEASSET_BALANCE: '123',
      TEST_MODE: 'true',
    };

    it('allows BTC/USDT', () => {
      expect.assertions(1);
      const config = checkConfigOptions(validLiveCEXdisabledConf);
      expect(config).toEqual({
        ...config,
        CEX: 'BINANCE',
        TEST_MODE: true,
        BASEASSET: 'BTC',
        QUOTEASSET: 'USDT',
        CEX_BASEASSET: 'BTC',
        CEX_QUOTEASSET: 'USDT',
      });
    });

    it('allows BTC/USDT with custom CEX_BASEASSET/CEX_QUOTEASSET', () => {
      expect.assertions(1);
      const config = checkConfigOptions({
        ...validLiveCEXdisabledConf,
        CEX_BASEASSET: 'XBT',
        CEX_QUOTEASSET: 'USD',
      });
      expect(config).toEqual({
        ...config,
        CEX: 'BINANCE',
        TEST_MODE: true,
        BASEASSET: 'BTC',
        QUOTEASSET: 'USDT',
        CEX_BASEASSET: 'XBT',
        CEX_QUOTEASSET: 'USD',
      });
    });

    it('allows ETH/BTC trading pair', () => {
      const config = checkConfigOptions({
        ...validLiveCEXdisabledConf,
        ...{
          CEX_BASEASSET: 'eth',
          CEX_QUOTEASSET: 'btc',
          BASEASSET: 'eth',
          QUOTEASSET: 'btc',
        },
      });
      expect(config).toEqual({
        ...config,
        CEX: 'BINANCE',
        TEST_MODE: true,
        BASEASSET: 'ETH',
        QUOTEASSET: 'BTC',
        CEX_BASEASSET: 'ETH',
        CEX_QUOTEASSET: 'BTC',
      });
    });

    it('requires TEST_CENTRALIZED_EXCHANGE_BASEASSET_BALANCE', () => {
      expect.assertions(1);
      const config = {
        ...validLiveCEXdisabledConf,
      };
      delete config.TEST_CENTRALIZED_EXCHANGE_BASEASSET_BALANCE;
      expect(() => {
        checkConfigOptions(config);
      }).toThrowErrorMatchingSnapshot();
    });

    it('requires TEST_CENTRALIZED_EXCHANGE_QUOTEASSET_BALANCE', () => {
      expect.assertions(1);
      const config = {
        ...validLiveCEXdisabledConf,
      };
      delete config.TEST_CENTRALIZED_EXCHANGE_QUOTEASSET_BALANCE;
      expect(() => {
        checkConfigOptions(config);
      }).toThrowErrorMatchingSnapshot();
    });
  });

  describe('TEST_MODE disabled', () => {
    const validLiveCEXenabledConf = {
      LOG_LEVEL: 'trace',
      CEX: 'kraken',
      CEX_API_KEY: '123',
      CEX_API_SECRET: 'abc',
      DATA_DIR: '/some/data/path',
      OPENDEX_CERT_PATH: '/some/cert/path',
      OPENDEX_RPC_HOST: 'localhost',
      OPENDEX_RPC_PORT: '1234',
      MARGIN: '0.06',
      BASEASSET: 'BTC',
      QUOTEASSET: 'USDT',
      TEST_MODE: 'false',
    };

    it('allows BTC/USDT', () => {
      expect.assertions(1);
      const config = checkConfigOptions(validLiveCEXenabledConf);
      expect(config).toEqual({
        ...config,
        CEX: 'KRAKEN',
        TEST_MODE: false,
      });
    });

    it('allows USDT/DAI trading pair', () => {
      const config = checkConfigOptions({
        ...validLiveCEXenabledConf,
        ...{
          CEX_BASEASSET: 'USDT',
          CEX_QUOTEASSET: 'DAI',
          BASEASSET: 'USDT',
          QUOTEASSET: 'DAI',
        },
      });
      expect(config).toEqual({
        ...config,
        CEX: 'KRAKEN',
        TEST_MODE: false,
        CEX_BASEASSET: 'USDT',
        CEX_QUOTEASSET: 'DAI',
        BASEASSET: 'USDT',
        QUOTEASSET: 'DAI',
      });
    });

    it('allows CEX[BTC/USD], OpenDEX[BTC/USDT]', () => {
      expect.assertions(1);
      const config = checkConfigOptions({
        ...validLiveCEXenabledConf,
        ...{
          CEX_BASEASSET: 'BTC',
          CEX_QUOTEASSET: 'USD',
          BASEASSET: 'BTC',
          QUOTEASSET: 'USDT',
        },
      });
      expect(config).toEqual({
        ...config,
        CEX: 'KRAKEN',
        TEST_MODE: false,
        CEX_BASEASSET: 'BTC',
        CEX_QUOTEASSET: 'USD',
        BASEASSET: 'BTC',
        QUOTEASSET: 'USDT',
      });
    });

    it('allows CEX[USD/DAI], OpenDEX[USDT/DAI]', () => {
      expect.assertions(1);
      const config = checkConfigOptions({
        ...validLiveCEXenabledConf,
        ...{
          CEX_BASEASSET: 'USD',
          CEX_QUOTEASSET: 'DAI',
          BASEASSET: 'USDT',
          QUOTEASSET: 'DAI',
        },
      });
      expect(config).toEqual({
        ...config,
        CEX: 'KRAKEN',
        TEST_MODE: false,
        CEX_BASEASSET: 'USD',
        CEX_QUOTEASSET: 'DAI',
        BASEASSET: 'USDT',
        QUOTEASSET: 'DAI',
      });
    });

    it('allows ETH/BTC trading pair', () => {
      const config = checkConfigOptions({
        ...validLiveCEXenabledConf,
        ...{
          CEX_BASEASSET: 'ETH',
          CEX_QUOTEASSET: 'BTC',
          BASEASSET: 'ETH',
          QUOTEASSET: 'BTC',
        },
      });
      expect(config).toEqual({
        ...config,
        CEX: 'KRAKEN',
        TEST_MODE: false,
      });
    });

    it('requires CEX_API_KEY', () => {
      expect.assertions(1);
      const config = {
        ...validLiveCEXenabledConf,
      };
      delete config.CEX_API_KEY;
      expect(() => {
        checkConfigOptions(config);
      }).toThrowErrorMatchingSnapshot();
    });

    it('requires CEX_API_SECRET', () => {
      expect.assertions(1);
      const config = {
        ...validLiveCEXenabledConf,
      };
      delete config.CEX_API_SECRET;
      expect(() => {
        checkConfigOptions(config);
      }).toThrowErrorMatchingSnapshot();
    });
  });
});
