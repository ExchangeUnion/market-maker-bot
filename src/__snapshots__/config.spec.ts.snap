// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkConfigOptions TEST_MODE disabled requires CEX_API_KEY 1`] = `"Incomplete configuration. Please add the following options to .env or as environment variables: CEX_API_KEY"`;

exports[`checkConfigOptions TEST_MODE disabled requires CEX_API_SECRET 1`] = `"Incomplete configuration. Please add the following options to .env or as environment variables: CEX_API_SECRET"`;

exports[`checkConfigOptions TEST_MODE enabled requires TEST_CENTRALIZED_EXCHANGE_BASEASSET_BALANCE 1`] = `"Incomplete configuration. Please add the following options to .env or as environment variables: TEST_CENTRALIZED_EXCHANGE_BASEASSET_BALANCE"`;

exports[`checkConfigOptions TEST_MODE enabled requires TEST_CENTRALIZED_EXCHANGE_QUOTEASSET_BALANCE 1`] = `"Incomplete configuration. Please add the following options to .env or as environment variables: TEST_CENTRALIZED_EXCHANGE_QUOTEASSET_BALANCE"`;
